db/cryptodata/decrypter.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cryptodata
+
+import (
+	"regexp"
+	"encoding/json"
+	"strings"
+)
+
+const cryptoPrefix = "$crypto$"
+
+var encryptedJSONRegex = regexp.MustCompile(`"encrypted"\s*:\s*"true"`)
+
+// DecryptFunc is function that decrypts arbitrary data
+type DecryptFunc func(inData []byte) (data [] byte, err error)
+
+// ArbitraryDecrypter is interface for decrypting groups of data
+type ArbitraryDecrypter interface {
+	// Decrypt decrypts input data using provided decrypting function for arbitrary data
+	Decrypt(inData []byte, decryptFunc DecryptFunc) (data []byte, err error)
+}
+
+// DecrypterJSON is ArbitraryDecrypter implementations that can decrypt JSON values
+type DecrypterJSON struct{}
+
+// Decrypt tries to decrypt JSON data that are encrypted based on "encrypted": "true" presence in it.
+// Then it parses data as JSON as tries to lookup all top-level values that begin with $crypto$ and decrypt them
+// using provided arbitrary decrypt function.
+func (DecrypterJSON) Decrypt(inData []byte, decryptFunc DecryptFunc) (data []byte, err error) {
+	data = inData
+
+	if !encryptedJSONRegex.Match(inData) {
+		return
+	}
+
+	var jsonData map[string]interface{}
+	err = json.Unmarshal(inData, &jsonData)
+	if err != nil {
+		return
+	}
+
+	for k, v := range jsonData {
+		var stringVal string
+		switch t := v.(type) {
+		case string:
+			stringVal = t
+		case []byte:
+			stringVal = string(t)
```

Comment:
You shouldn't convert from `[]byte` to `string` if it's not needed. 

Here you are actually doing it multiple times. Once here to `string` and then below back to `[]byte`, that is not very efficient. You can use `bytes` package for working with bytes.

Actually.. reading the docs for `json.Unmarshal` it seems the `[]byte` will never be used for any type of data in json. So scratch my previous suggestion to add it here.

However what might be an issue is nested data. Since you are only going over top level fields, this won't work for json data like this:

```json
{
  "encrypted": true,
  "nested": {
    "abc": "aaa",
    "zxy": "$crypto$xxxxxxxx"
  }
}
```

You should probably do some kind of loop that will check if the type is `map[string]interface{}` and if so then do another loop to check for encrypted fields.